Sources/NIO/CircularBuffer.swift

```diff
@@ -28,61 +28,69 @@ public struct CircularBuffer<Element>: CustomStringConvertible, AppendableCollec
     private var buffer: ContiguousArray<Element?>
 
     /// The index into the buffer of the first item
-    private(set) /* private but tests */ internal var headIdx = 0
+    private(set) /* private but tests */ internal var headIdx: Index
 
     /// The index into the buffer of the next free slot
-    private(set) /* private but tests */ internal var tailIdx = 0
+    private(set) /* private but tests */ internal var tailIdx: Index
 
     public struct Index: Strideable, Comparable {
-        @usableFromInline
-        var _value: Int
-
-        public typealias Stride = Int
+        @usableFromInline var _value: ContiguousArray<Element?>.Index
+        @usableFromInline var _headValue: ContiguousArray<Element?>.Index
+        @usableFromInline let underlyingCount: Int
+        @usableFromInline var mask: Int { return self.underlyingCount - 1 }
 
         @inlinable
-        internal init(_value: Int) {
-            self._value = _value
+        internal init(value: ContiguousArray<Element?>.Index, headValue: ContiguousArray<Element?>.Index, underlyingCount: Int) {
+            self._headValue = headValue
+            self.underlyingCount = underlyingCount
+            self._value = (headValue + value) & (underlyingCount - 1)
         }
 
         @inlinable
         public func distance(to other: Index) -> Int {
-            return self._value.distance(to: other._value)
+            let head = self._headValue
+            if self._value >= head && other._value >= head {
+                return self._value.distance(to: other._value)
+            } else if self._value >= head && other._value <= head {
+                return self._value.distance(to: underlyingCount) + other._value
+            } else if self._value <= head && other._value >= head {
+                return self._value.distance(to: 0) + underlyingCount.distance(to: other._value)
+            } else {
+                return self._value.distance(to: other._value)
+            }
         }
 
         @inlinable
         public func advanced(by n: Int) -> Index {
-            return Index(_value: self._value + n)
+            var index = Index(value: 0, headValue: self._headValue, underlyingCount: self.underlyingCount)
+            index._value = (self._value + n) & (self.mask)
+            return index
         }
 
         @inlinable
         public static func < (lhs: Index, rhs: Index) -> Bool {
-            return lhs._value < rhs._value
-        }
-
-        @inlinable
-        public static func <= (lhs: Index, rhs: Index) -> Bool {
-            return lhs._value <= rhs._value
-        }
-
-        @inlinable
-        public static func == (lhs: Index, rhs: Index) -> Bool {
-            return lhs._value == rhs._value
+            let head = lhs._headValue
+            if lhs._value >= head && rhs._value >= head {
```

Comment:
I think it might be worth adding a few test cases to hit all those cases

Reply:
A bit off topic but I'm unable to run the tests in XCode with code coverage enabled so I can't see if all code paths are covered already by other tests. Is this a general issue or is it just me?

Reply:
@JovanMilenkovic that's unfortunately a bug in the Xcode beta...